Feature Firebase

yanevdw opened the pull request:
Implemented and integrated Firebase and authentication. 

This includes:
- Authenticating users with existing users when they attempt to log in.
- Creating new users and adding them to the database when they register.
- Persisting the user using an observable and signal to keep track of the logged-in user.

yanevdw committed:
feat: :sparkles: Added firebase

Added firebase to project

yanevdw committed:
feat: :recycle: Updated environments configuration

Updated environments location

yanevdw committed:
feat: :sparkles: Added firebase authentication

Added firebase authentication for users and created functions for to authenticate users when they log in as well as register new users

yanevdw committed:
feat: :recycle: Updated name of form input and added feedback for user to see that log in was successfull

yanevdw committed:
feat: :sparkles: Added logged in user's details to header

Removed hard-coded user name and initials and used current (logged in) user's details.

yanevdw committed:
feat: :sparkles: Added a service for all the interactions with the database.

Created a service to handle all interactions with the Firestore database and added some functions to access collections in the database.

yanevdw committed:
feat: :sparkles: Added Firestore

Added Firestore for database integration.

yanevdw committed:
feat: :sparkles: Added some database/firestore integration to the collections page.

Began implementation of the collections component by getting the id of the user's bookshelf from the bookshelf table in the database.

yanevdw committed:
feat: :sparkles: Added interfaces

Added interfaces for the Bookshelf and Book collections

yanevdw committed:
feat:  :recycle: Updated routing implementation for login and register

Used RouterLink instead of functions to navigate back to login/register from the register/login and removed unused code.

yanevdw committed:
feat: :bug: Fixed routing typo

Added forward slashes needed to route to the correct page.

yanevdw committed:
feat: :sparkles: Added a cookie

Added a cookie for the userId.

yanevdw committed:
feat: :sparkles: Added a logout function

Added a logout function which also deletes the cookie.

yanevdw committed:
feat: :sparkles: Added integration to create a bookshelf for a new user.

Made use of the data service to call the createBookshelf function when a new user is created to automatically create a bookshelf for every new user and to ensure every user has a bookshelf.

yanevdw committed:
feat: :recycle: Added cookie expiration

Added functionality which allows cookies to expire after 12 hours.

yanevdw committed:
feat: :recycle: Moved data service to correct directory

Moved data service to services directory.

yanevdw committed:
feat: :sparkles: Added functionality to create a bookshelf.

Added a function which creates a bookshelf for a user which will only be used when a new user is created.

yanevdw committed:
feat: :sparkles: Added functionality to get the saved cookie.

Added a function that gets the current user's id from the cookie.

yanevdw committed:
feat: :bug: Fixed unused constant

Used constant instead of document.cookie again.

yanevdw committed:
feat: :heavy_plus_sign: Added ng-icons

Added ng-icons to incorporate icons.

yanevdw committed:
feat: :recycle: Updated implementation to observables

Switched implementation of auth from promises to observables

yanevdw committed:
feat: :recycle: Updated implementation where the curren user's id is required.

Updated implementation of functions where userId is required to get it from the cookie.

yanevdw committed:
feat: :sparkles: Added log out button and functionality

Added a button for users to log out and implemented the functionality to call the logout function to sign out the user.

yanevdw committed:
feat: :sparkles: Added subscriptions.

Added subscriptions for login and register.

yanevdw committed:
feat: :recycle: Updated cookie implementation and authentication

Added a new cookie for the user name, edited cookie functions and added authentication state checker to keep track of current user.

yanevdw committed:
feat: :fire: Removed unused utils directory

Removed directory and files no longer required or in use.

yanevdw committed:
feat: :sparkles: Added new interface for the user

Added UserInfo interface that represents the current user.

yanevdw committed:
feat: :recycle: Added a signal instead of using cookies

Used a signal to persist the current/logged in user.

Julianna-Venter commented on src/app/components/collections/collections.component.ts:
```diff
@@ -9,4 +11,32 @@ import { RouterLink } from '@angular/router';
+    });
+
+    const bookshelfResult$ = this.dataService.getBookshelf(
+      this.loggedInUserId!,
```
maybe just add `?? ''` for this since it is not guaranteed that loggedInUserId will be set when you do this, or add an if statement check

Julianna-Venter commented on src/app/components/collections/collections.component.ts:
```diff
@@ -9,4 +11,32 @@ import { RouterLink } from '@angular/router';
+      this.loggedInUserId!,
+    );
+    let bookshelfSubscription = bookshelfResult$.subscribe((bookshelf) => {
+      this.bookshelfId = bookshelf[0].id;
```
`bookshelf?[0]?.id`

yanevdw replied:
Updated to `bookshelf?.[0]?.id`.

Julianna-Venter commented on src/app/components/collections/collections.component.ts:
```diff
@@ -9,4 +11,32 @@ import { RouterLink } from '@angular/router';
+  }
+
+  ngOnDestroy(): void {
+    this.authService.isUserSet$.unsubscribe();
```
a lot of this will change when you migrate to NgRx I expect, but just a reminder that there is the self destruct pipe piece thing that you can add to the subscription if you don't want to do the ngOnInit ngOnDestroy thing. Just a suggestion if you want to look into it. It likely won't even be needed once you add NgRx.

yanevdw replied:
Thank you, will keep this in mind.

Julianna-Venter commented on src/app/components/header/header.component.html:
```diff
@@ -4,11 +4,21 @@
-      Y
-    </p>
-    <p class="font-tilt-neon font-semibold text-lg text-pale-brown">Hi Yané</p>
+    @if (loggedInUser !== null && loggedInUser !== "" && loggedInUser !== undefined) {
```
this is still from when we were testing, but you can change this back to `loggedInUser ? loggedInUser : ''`, the whole if statement is a bit verbose, but not wrong

yanevdw committed:
feat: :fire: Removed unused code

Removed code that has been commented out.

Julianna-Venter commented on src/app/components/header/header.component.ts:
```diff
@@ -1,12 +1,41 @@
+    this.authService.isUserSet$.subscribe((user) => {
+      if (user) {
+        this.loggedInUser = this.authService.currentUserSignal()?.name;
+      } else {
```
else statement is empty, you can either remove it or have a sane default for loggedInUser if it fails to set the isUSerSet flag

yanevdw replied:
Thank you, I accidentally overlooked this.

Julianna-Venter commented on src/app/components/register/register.component.html:
```diff
@@ -55,38 +55,42 @@ <h2 class="font-pacifico w-full font-semibold text-3xl text-amaranth">
+              Please confirm your password!
+            }
+            @if (control.errors?.['confirm']) {
+              Two passwords that you enter is
```
good engles

yanevdw replied:
Updated, this was still from the UI library.

Julianna-Venter commented on src/app/guards/auth.guard.ts:
```diff
@@ -1,13 +1,42 @@
+  //   authService.isUserSet$.next(true);
+  //   return true;
+  // }
+  // // return true;
```
ist thou going to use thine code hence forth? negatory? banish it.

Julianna-Venter replied:
nevermind...

yanevdw replied:
Has been removed in latest commit.

Julianna-Venter commented on src/app/guards/auth.guard.ts:
```diff
@@ -1,13 +1,42 @@
+      router.navigate(['/login']);
+      return false;
+    } else {
+      // router.navigate(['/']);
```
a repeat blunder (I know I wrote it but don't trust me)

yanevdw replied:
Removed in latest commit.

Julianna-Venter commented on src/app/models/states.d.ts:
```diff
@@ -0,0 +1,27 @@
+  moods: string[];
+  pace: string;
+  pages: number;
+  pages_read?: number;
```
are you going to have the user track this info manually?

yanevdw replied:
Yes,  this will appear when the book is in the "currently read" category. Allows the user to track reading progress.

Julianna-Venter commented on src/app/services/auth.service.ts:
```diff
@@ -1,16 +1,53 @@
+      signInWithEmailAndPassword(
+        this.firebaseAuth,
+        email.trim(),
+        password.trim(),
```
smart to trim it, I forgot about this...👏

Julianna-Venter reviewed (approved):
Approved. Minor comments here and there, nothing game changing. If you make these changes then all will be well.

yanevdw committed:
feat: :recycle: Fixed minor bugs and added minor updates.

Updated implementation to remove testing and improve structure and added fixes based on feedback.

Meggie-H reviewed (approved):
I can't see anything wrong that stands out to me. Well done
